File: rust/quickq-core/src/dlq.rs
use crate::error::Result;
use crate::job::Job;
use crate::storage::sqlite::{DeadJob, SqliteStorage};

/// Dead letter queue manager.
pub struct DeadLetterQueue {
    storage: SqliteStorage,
}

impl DeadLetterQueue {
    pub fn new(storage: SqliteStorage) -> Self {
        Self { storage }
    }

    /// Move a failed job to the dead letter queue.
    pub fn move_to_dlq(&self, job: &Job, error: &str, metadata: Option<&str>) -> Result<()> {
        self.storage.move_to_dlq(job, error, metadata)
    }

    /// List dead letter entries.
    pub fn list(&self, limit: i64, offset: i64) -> Result<Vec<DeadJob>> {
        self.storage.list_dead(limit, offset)
    }

    /// Re-enqueue a dead letter job. Returns the new job ID.
    pub fn retry(&self, dead_id: &str) -> Result<String> {
        self.storage.retry_dead(dead_id)
    }

    /// Purge dead letter entries older than the given number of milliseconds ago.
    pub fn purge(&self, older_than_ms: i64) -> Result<u64> {
        self.storage.purge_dead(older_than_ms)
    }
}

File: sdks/node/src/cli/commands/dlq.ts
import type { Command } from "commander";
import { connect, type GlobalOptions } from "../connect";
import { printJson, printTable } from "../output";

export function registerDlq(program: Command): void {
  const dlq = program.command("dlq").description("Dead-letter queue operations");

  dlq
    .command("list")
    .description("List dead-letter entries")
    .option("--limit <n>", "page size", "50")
    .option("--offset <n>", "page offset", "0")
    .action((options: { limit?: string; offset?: string }, command: Command) => {
      const globals = command.optsWithGlobals() as GlobalOptions;
      const queue = connect(globals);
      const dead = queue.deadLetters(
        options.limit ? Number(options.limit) : undefined,
        options.offset ? Number(options.offset) : undefined,
      );
      if (globals.json) {
        printJson(dead);
        return;
      }
      printTable(
        dead.map((entry) => ({
          id: entry.id,
          task: entry.taskName,
          queue: entry.queue,
          error: entry.error ?? "",
        })),
      );
    });

  dlq
    .command("retry <deadId>")
    .description("Re-enqueue a dead-letter entry")
    .action((deadId: string, _options: unknown, command: Command) => {
      const queue = connect(command.optsWithGlobals() as GlobalOptions);
      printJson({ id: queue.retryDead(deadId) });
    });

  dlq
    .command("delete <deadId>")
    .description("Delete a dead-letter entry")
    .action((deadId: string, _options: unknown, command: Command) => {
      const queue = connect(command.optsWithGlobals() as GlobalOptions);
      printJson({ deleted: queue.deleteDead(deadId) });
    });

  dlq
    .command("purge")
    .description("Purge dead-letter entries older than a cutoff")
    .option("--older-than-ms <n>", "age cutoff in ms", "0")
    .action((options: { olderThanMs?: string }, command: Command) => {
      const queue = connect(command.optsWithGlobals() as GlobalOptions);
      printJson({ purged: queue.purgeDead(Number(options.olderThanMs ?? 0)) });
    });
}
